[pwrmgr] Transferring document contents to markdown

tjaychen opened the pull request:
Signed-off-by: Timothy Chen <timothytim@google.com>

tjaychen commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,231 @@
+
+## Register Table
+
+{{< registers "hw/top_earlgrey/ip/pwrmgr/data/autogen/pwrmgr.hjson" >}}
```
@eunchan 
@msfschaffner 
@gkelly 

do you guys know of a better way for this to be handled? It feels like the documentation we generate should be to a specific top, and in the documentation we should be able to just say `TOP_VERSION` or something. 

What do you guys think?

gkelly replied:
I think that's ideal, because right now we usually have to hunt for the configuration that blocks are instantiated with inside of the toplevel when trying to set things up. I think the issue is that we need to make that generation happen at the `build_docs.py` level, and then populate it down through the documentation.

tjaychen commented:
Adding @mcy since he helped review the original google doc.

tjaychen commented:
Need to link the diagrams, hopefully will do that today.

tjaychen commented:
let me know if this markdown looks good overall.
i've not transferred the programmers model portion in yet because there are still LOTS of comments on that one, and i want to make sure others have a chance to chime in.

rswarbrick commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+
+### Reset Recording
+
+Of the resets directly handled by the power manager (peripheral reset request, escalation and low power entry), the power manager sequences only 1 at a time (see state diagrams).
```
I think this would be easier to read as something like "The power manager only handles one reset request at once, which means it serializes any overlapping reset requests that it must handle directly (peripheral reset request, escalation and low power entry)."

Also, I looked at the state diagrams and I'm not sure how they support the statement. Is the point that there's just one "reset_req" pathway? I'm not sure whether the text needs a pointer to the state diagrams but, if it keeps the pointer, it probably needs to explain why they demonstrate what it's saying.

tjaychen replied:
i completely reworded this section. Can you let me know if it's clearer now?

rswarbrick replied:
Yep, I think I understand now. Thanks!

rswarbrick commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+### Reset Recording
+
+Of the resets directly handled by the power manager (peripheral reset request, escalation and low power entry), the power manager sequences only 1 at a time (see state diagrams).
+This means if all reset requests happen at approximately the same time, the power manager will handle in a first come first serve manner.
```
Should probably be "will handle them"

rswarbrick commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+
+Of the resets directly handled by the power manager (peripheral reset request, escalation and low power entry), the power manager sequences only 1 at a time (see state diagrams).
+This means if all reset requests happen at approximately the same time, the power manager will handle in a first come first serve manner.
+When the handling of the first is completed, it observes there is another pending request and handles accordingly.
```
Here "it" is a little unclear, because there's been another object since. Probably "... is completed, the power manager observes ..." is better.

rswarbrick commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+This means if all reset requests happen at approximately the same time, the power manager will handle in a first come first serve manner.
+When the handling of the first is completed, it observes there is another pending request and handles accordingly.
+This is done because low power resets and peripheral requested resets lead to different behaviors.
+When the power manager commits to handling a specific request, it sends a reset hint to the reset manager to inform it of what reasons to record should the processor reset.
```
This feels a little hard to parse. Maybe something like this would be better? "... it sends a reset hint to the reset manager, informing it what reasons to record if it resets the processor."

tjaychen replied:
i reworded this

rswarbrick commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+This is done because low power resets and peripheral requested resets lead to different behaviors.
+When the power manager commits to handling a specific request, it sends a reset hint to the reset manager to inform it of what reasons to record should the processor reset.
+
+If for example a low power entry request is seen first by the power manager, it transitions the system to a low power state and informs the reset manager to record “low power exit” as the reset reason.
```
This isn't quite right because "it" sounds like it's the request, not the power manager. The simplest fix is rewriting in the active voice: "If for example the power manager first sees a low power entry request, ..." (and I'd put "for example" in commas, but I think that's a matter of preference)

tjaychen replied:
i reworded this

rswarbrick commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+If during this transition a watchdog request is observed, the power manager then uses the watchdog as a wakeup cause to resume the system and handle the associated request.
+As this request is handled, the reset manager records a second reset reason of “watchdog”.
+
+Ultimately when control is returned to software, it sees two reset reasons and must handle accordingly.
```
Probably "and must handle them accordingly"? Otherwise "handle" has no object.

tjaychen replied:
fixed

rswarbrick commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+
+## Wakeup Recording
+
+Similar to [reset handling](#Reset-Request-Handling), wakeup signals are only observed during slow FSM low power; however its recording is continuous until explicitly disabled by software.
```
Do you mean the recording of wakeup signals is continuous? Then I think you need "however their recording is continuous".

tjaychen replied:
fixed

rswarbrick commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+Similar to [reset handling](#Reset-Request-Handling), wakeup signals are only observed during slow FSM low power; however its recording is continuous until explicitly disabled by software.
+
+The wakeup recording begins when the fast FSM transitions out of ACTIVE state and continues until explicitly disabled by software.
+This ensures wakeup recording is continuous until software has set up the appropriate peripherals to handle future events.
```
This last sentence feels a little circular ("the recording is continuous to ensure that it's continuous"). Presumably the point is to avoid missing anything, so maybe "This ensures that no wakeup events are missed by coming in before software has set up ..." ?

tjaychen replied:
fixed

rswarbrick commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+
+### Power-Up Handling
+
+The AST automatically takes flash out of power down state as part of power manager's power up request.
```
I think you've been using "the power manager" elsewhere in this document (which I like!). Probably a missing "the" for consistency here.

tjaychen replied:
fixed

rswarbrick commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+    *   If in flash, this information is unscrambled
+    *   This item is TBD and must be further discussed.
+*   Program redundancy or configuration information into the flash controller and lock it from further modification
+*   Initialize flash scrambling and programming any excluded regions based on OTP or other information.
```
Probably "and program" here? Or "Initializing", but that wouldn't be consistent with the previous bullet points.

tjaychen replied:
fixed

rswarbrick commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+### Power-Down Handling
+
+When the device enters deep sleep, the flash is automatically put into power down mode by the AST.
+When the device exits low power state, the flash is automatically powered up and it is boot ROM responsibility to poll for flash power-up complete.
```
This should probably be "boot ROM's responsibility", but I'd probably rewrite as "the responsibility of boot ROM" because it's a bit easier to read.

tjaychen replied:
fixed

rswarbrick commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+### Deep Sleep or Standby
+
+This represents the lowest power mode of the device (outside of full power down or device held in reset).
+During this state, the following states of the device shall be true
```
I'm not sure what it means for a state to be true. Probably this could just be "During this state:".

tjaychen replied:
fixed!

rswarbrick commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+This represents the lowest power mode of the device (outside of full power down or device held in reset).
+During this state, the following states of the device shall be true
+
+*   All clocks other than the KHz slow clock shall be turned off at the source.
+*   All non-always-on digital domains are powered off
```
One of these lines reads like an RFC ("the client shall do XYZ") and the other sounds more informative. I don't think the distinction is intentional? I'd probably change to the second phrasing, so "clocks ... are turned off", unless there's a good reason not to.

tjaychen replied:
thanks for catching this, fixed!

rswarbrick commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+*   All clocks other than the KHz slow clock shall be turned off at the source.
+*   All non-always-on digital domains are powered off
+*   I/O power domains may or may not be off
+    *   The state of the IO power domain shall have no impact on the digital core’s power budget
```
Another "shall". What is required to obey this restriction? (It's a genuine question - I might have just missed something). Or is this supposed to be saying that circuitry in the IO power domain doesn't count towards the digital core's power budget? In which case, I think I'd write so explicitly.

tjaychen replied:
ah yeah i can clarify.  It's more that the pad designs have to be done right such that a floating IO supply does not cause extra leakage on the digital domain.

rswarbrick commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+This represents a fast low power mode of the device that compromises between power consumed and resume latency.
+During this state, the following states of the device shall be true
+
+*   All clocks other than the KHz slow clock shall be turned off at the source (AST)
```
As above with the "shall"s, but also this has an "(AST)" suffix. Probably should be consistent with the previous list?

tjaychen replied:
fixed!

rswarbrick commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+During this state, the following states of the device shall be true
+
+*   All clocks other than the KHz slow clock shall be turned off at the source (AST)
+*   All non-always-on digital domains are kept on for fast resume
```
If I understand correctly, any other digital domain is always-on, so is also kept on. Maybe this could be simplified to just "All digital domains..."?

tjaychen replied:
good catch, i've fixed this to say all power domains.

tjaychen commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+Low power entry is always initiated by software, while low power exit is always initiated by a previously setup hardware event such as pins or internal timers.
+The power manager processes the software and hardware requests to perform the appropriate actions.
+
+Reset scenarios refer to non-por events that cause the device to reboot.
```
> Hi Tim! Thanks for writing this up, and for making a PR that has given me a prod to read through it in detail. The design looks really sensible to me.
> 
> The bajillion comments below are all grammar and style nits that I noticed as I went through: probably worth fixing, but the large number is a function of document length. It doesn't mean that I think this is terrible :-)

Thank you for the detailed review @rswarbrick !

tjaychen replied:
fixed!

tjaychen commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+
+## Description
+
+The power manager sequeces power, clocks, and reset resources of the design through cold boot, low power entry/exit and reset scenarios.
```
fixed!

tjaychen commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+
+Low power entry is the process in which the device enters one of two low power modes (sleep or deep sleep).
+Low power exit is the process in which the device exits low power mode and returns to active state.
+Low power entry is always initiated by software, while low power exit is always initiated by a previously setup hardware event such as pins or internal timers.
```
makes sense, fixed!

tjaychen commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+The power manager processes the software and hardware requests to perform the appropriate actions.
+
+Reset scenarios refer to non-por events that cause the device to reboot.
+There are various stimuls that can cause such a reset, ranging from exteranl user input to watchdog.
```
fixed!

tjaychen commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+
+# Theory Of Operation
+
+The power manager controls power domain(s) turn on / turn off, the root resets in conjunction with the reset manager, the root clock enables in conjunction with AST / clock manager and various power up activities such as OTP sensing and life cycle initiation before allowing software to execute.
```
yes, thanks for pointing all this out. I have converted it to a list and it does read much more easily.

tjaychen commented:
could i get an approve from someone to push this document in?

imphil commented:
@tjaychen there are a couple more comments from @rswarbrick unresolved; you might need to click on the "show all comments"-like link (or go to the "Files changed" tab). Or are those comments not relevant any more? (I didn't check all of them, some felt still relevant.)

sriyerg commented on hw/ip/pwrmgr/data/pwrmgr.prj.hjson:
```diff
@@ -0,0 +1,19 @@
+        version:            "0.1",
+        life_stage:         "L1",
+        design_stage:       "D1",
+        verification_stage: "V1",
```
`pwrmgr` is not verified at the IP level. Wonder if this should remain at V0.

tjaychen commented:
> @tjaychen there are a couple more comments from @rswarbrick unresolved; you might need to click on the "show all comments"-like link (or go to the "Files changed" tab). Or are those comments not relevant any more? (I didn't check all of them, some felt still relevant.)

thanks, i completely missed those.

tjaychen commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+
+## Overall Sequencing
+
+The power manager primarily is composed of two state machines.
```
yeah i dont know what i was trying to either... i should have just said the power manager is composed of two state machines...

tjaychen commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+The following diagram breaks down the general functionality of both.
+
+
+![Power Manager FSMs](pwrmgr_fsms.svg)
```
i added some text to explain the colors, and yes i do think I was supposed to have bolded Active and not dis_clks.
Maybe I accidentally copy / pasted and forgot.  Let me just remove it for now.

rswarbrick replied:
Oh, sorry, I meant the colours in the text (so "ack_pwrup" is green, "!req_pwrup" is orange). Is this just to show that the signal comes from the other FSM?

Assuming that "!foo" means "foo is not asserted", it might be sensible to change it to "foo = 0", otherwise you collide with the send/receive notation that's usually used for communicating state machines (for example, see the Wikipedia article on [CSP](https://en.wikipedia.org/wiki/Communicating_sequential_processes)): I originally thought that this is what you meant.

I think I still don't understand the bold ellipses and signals in the FSM diagram either.

And (nit) it would still be nice if the dashed line was extended to the bottom or vertically centred (I'm listing it again so that everything appears in the last comment). Obviously, this isn't massively important!

tjaychen replied:
oops! i actually fixed some of this but did not upload the diagram. (also forgot to extend the dashed line).
I'll do that soon and let you know.

tjaychen commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+The slow clock domain FSM (referred to as the slow FSM from here on) resets to the RESET state.
+This state is released by por\_rst\_n, which is supplied from the reset controller.
+Por\_rst\_n releases when the reset controller detects the root power domains (aon_pok) of the system are ready.
+Please see the [AST interface]() for more details.
```
ah this is because the AST documentation PR has not been merged yet.  Once it is, I'm going to reference there.
I didn't want to put the link down yet because it would be broken. (although I guess it's broken now, so let me just remove it and re-introduce it later).

rswarbrick replied:
Cool, just wanted to check this wasn't a typo.

tjaychen commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+
+Once the fast FSM acknowledges the power-up completion, the slow FSM transitions to IDLE and waits for a power down request.
+When a power down request is received, the slow FSM turns off AST clocks and power as directed by software configuration.
+This means the clocks and power are not always turned off, but rather controlled by software configurations in {{< regref "CONTROL" >}} prior to low power entry .
```
i prefer are rather 👍

tjaychen commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+The fast clock domain FSM (referred to as fast FSM from here on) resets to LOW\_POWER state and waits for a power-up request from the slow FSM.
+
+Once received, the fast FSM releases the life cycle reset stage (see [reset controller](https://docs.google.com/document/d/1oprdDwbm-_opDwMuu-kmmaFVwt4-f1EGqCf0nkO8VpM/edit?usp=sharing) for more details).
+This allows the OTP to begin sensing.
+Once OTP sensing completes (see [OTP controller](https://docs.google.com/document/d/1ga4_cN1Md9yYUANwh_1KdhlDdkp1gQxbUArwm8h1Rvw/edit?usp=sharing) for more details), the life cycle controller is initialized.
```
this is also the same thing as AST, PR not yet merged :)
I think i should take this out too until the PR lands

tjaychen commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+Once OTP sensing completes (see [OTP controller](https://docs.google.com/document/d/1ga4_cN1Md9yYUANwh_1KdhlDdkp1gQxbUArwm8h1Rvw/edit?usp=sharing) for more details), the life cycle controller is initialized.
+The initialization of the life cycle controller puts the device into its allowed operating state (see [life cycle controller](https://docs.google.com/document/d/1H6jHtX2xYgy3nmkWKeHMkYcP0vsPq9duyUtLu5imX2E/edit?usp=sharing) for more details).
+
+Once life cycle initialization is done, the fast FSM enables all second level clock gating (see clock controller for more details) and initiates strap sampling.
```
i think this one i can link the google doc for now (it's not sensitive like ast / otp, where i want to wait for the PR)

tjaychen commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+This separation is because the non-debug reset does not affect the life cycle controller, non-volatile storage controllers and alert states.
+There is thus no need to sequence its operation like the others.
+
+The power controller observes reset requests only in two states - the slow FSM LOW\_POWER state and the fast FSM ACTIVE state - these are considered static states of the system.
```
i think i'll put the state names in back ticks for now, hopefully that makes a clearer separation.  They are directly related to the design, so it seems reasonable to me.

rswarbrick replied:
Looks sensible to me, thanks.

tjaychen commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+
+The power manager controls power domain(s) turn on / turn off, the root resets in conjunction with the reset manager, the root clock enables in conjunction with AST / clock manager and various power up activities such as OTP sensing and life cycle initiation before allowing software to execute.
+
+In the future, if there is a need to perform ROM hashing as an integrity check, that will also be managed by the power manager.
```
fixed!

tjaychen commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+## Block Diagram
+
+See the below high level block diagram that illustrates the connections between the power manager and various system components.
+Blocks outlined with a solid magenta line are always on; while blocks outlined with a dash magenta line are a mix of always-on and non-always-on components.
```
fixed!

tjaychen commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+
+
+*   Slow FSM IDLE and fast FSM ACTIVE
+*   Slow FSM LOW\_POWER and fast FSM LOW\_POWER
```
fixed!

tjaychen commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+
+The slow clock domain FSM (referred to as the slow FSM from here on) resets to the RESET state.
+This state is released by por\_rst\_n, which is supplied from the reset controller.
+Por\_rst\_n releases when the reset controller detects the root power domains (aon_pok) of the system are ready.
```
fixed!

tjaychen commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+A low power request is initiated by software through a combination of WFI and software low power hint in {{< regref "CONTROL" >}}.
+Specifically, this means if software issues only WFI, the power manager does not treat it as a power down request.
+The notion of WFI is exported from the processor.
+For Ibex, this is currently in the form of core\_sleeping\_o.
```
fixed!

tjaychen commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+## Reset Request Handling
+
+There are 3 reset requests in the system - peripheral requested reset such as watchdog, alert escalation reset and non-debug module reset.
+The flash brownout is handled separately and described in [flash handling section](#Flash-Handling) below.
```
fixed!

tjaychen commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+There are 3 reset requests in the system - peripheral requested reset such as watchdog, alert escalation reset and non-debug module reset.
+The flash brownout is handled separately and described in [flash handling section](#Flash-Handling) below.
+
+The watchdog and alert escalation resets are handled directly by the power manager, while the non-debug module reset is handled by the reset controller.
```
fixed!

tjaychen commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+There is thus no need to sequence its operation like the others.
+
+The power controller observes reset requests only in two states - the slow FSM LOW\_POWER state and the fast FSM ACTIVE state - these are considered static states of the system.
+When a reset request is received during slow FSM LOW\_POWER state, the system begins its usual power up sequence even if a wakeup has not been received.
```
fixed!

tjaychen commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+
+### Reset Requests Received During Other States
+
+All other states in the slow / fast FSM are considered transition states.
```
fixed!

tjaychen commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,233 @@
+
+All other states in the slow / fast FSM are considered transition states.
+Resets are not observed in other states because the system will always be transitioning towards one of the steady states (the system is in the process of powering down or powering up).
+Once a steady state is reached, the reset request is then observed and action taken.
```
fixed!

rswarbrick commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,256 @@
+
+The following diagram breaks down the general functionality of both.
+The state machines are colored based on their clock domains.
+The green state machine is clocked by the normal fixed domain, while the organge state machine is clocked by the slow domain.
```
Typo: "organge"

tjaychen replied:
i was obviously going for orangutan there...

rswarbrick commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,256 @@
+
+
+*   Slow FSM `Idle` and fast FSM `Active`
+*   Slow FSM `Low_Power` and fast FSM `Low_power`
+
+The slow FSM `Low_power` and fast FSM `Active` states specifically are concepts useful when examining [reset handling](#Reset-Request-Handling).
```
Sorry for the nitpicks, but I don't think the capitalisation is consistent for `Low_power` / `Low_Power` here. Also, do you need the underscores? I think you've got spaces in the diagram.

tjaychen replied:
good catch, i didnt even notice some were capitalized and some were not.

rswarbrick commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,256 @@
+
+When a reset request is received during fast FSM `Active` state, the fast FSM asserts resets and transitions back to its `Low_power` state.
+The normal power-up process described [above](#Fast-Clock-Domain-FSM) is then followed to release the resets.
+Note in this case, the slow FSM is “not activated” and remains in its `Idle` state.
```
I think we avoid non-ASCII characters (smart quotes).

tjaychen replied:
thanks!

rswarbrick commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,256 @@
+- Direct reset requests by peripherals or alert escalation
+
+The power manager does not handle the non-debug-module request (please see reset controller).
+For the remaining two reset causes, the power manager handles only 1 pathway at a time (see state diagrams).
```
Ah! Now I understand! Could you note in the text that "low power entry" corresponds to the `sleep_req` branch? It took me a few seconds to figure this out.

tjaychen replied:
added

rswarbrick commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,256 @@
+
+The power manager does not handle the non-debug-module request (please see reset controller).
+For the remaining two reset causes, the power manager handles only 1 pathway at a time (see state diagrams).
+This means if reset request and low power entry collide, the power manager will handle on a first come first serve basis.
```
Probably "will handle **them** on" and "first come, first serve**d**"

tjaychen replied:
fixed

rswarbrick commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,256 @@
+This is done because low power resets and peripheral requested resets lead to different behaviors.
+When the power manager commits to handling a specific request, it informs the reset manager why it has reset the processor.
+
+For example, assume low power entry request arrives slightly ahead of reset requests, the power manager will:
```
I think this needs an article "assume **a** low power entry request...". Also (nit), you probably want to break the sentence to give something like "... reset requests. Then the power manager will:". (Otherwise, the comma separates two primary clauses, which ain't allowed)

tjaychen replied:
fixed

rswarbrick commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,256 @@
+
+For example, assume low power entry request arrives slightly ahead of reset requests, the power manager will:
+- Transition the system into low power state.
+- Inform reset manager to record "low power exit" as the reset reason.
```
Here and two lines lower, I think you want "**the** reset manager" (This matches e.g. line 153 and I think it's just nicer English!)

tjaychen replied:
fixed

rswarbrick commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,256 @@
+For example, assume low power entry request arrives slightly ahead of reset requests, the power manager will:
+- Transition the system into low power state.
+- Inform reset manager to record "low power exit" as the reset reason.
+- Once in low state, transition the system to Active state by using the reset request as a wakeup indicator.
```
Do you want back ticks on "Active" as above?

tjaychen replied:
fixed!

rswarbrick commented on hw/ip/pwrmgr/doc/_index.md:
```diff
@@ -0,0 +1,256 @@
+- Inform reset manager to record "low power exit" as the reset reason.
+- Once in low state, transition the system to Active state by using the reset request as a wakeup indicator.
+- Inform reset manager to also record the peripheral that requested reset.
+- Once in active state, reset the system and begin normal power-up routines again.
```
I think this line should probably use the same formatting as line 158 (probably `Active` with backticks?)

tjaychen replied:
fixed!

rswarbrick reviewed (commented):
Hi Tim! Thanks for addressing all the comments. GitHub's decision to hide things is pretty irritating!

I've got a couple of other nits and questions, but this is much easier to read, thanks. Because of GitHub's helpful formatting, there's one outstanding question in the original list which gets hidden when you reload the page. To save you having to dig around, here's a [link](https://github.com/lowRISC/opentitan/pull/3119#discussion_r473685195).

tjaychen commented:
> Hi Tim! Thanks for addressing all the comments. GitHub's decision to hide things is pretty irritating!
> 
> I've got a couple of other nits and questions, but this is much easier to read, thanks. Because of GitHub's helpful formatting, there's one outstanding question in the original list which gets hidden when you reload the page. To save you having to dig around, here's a [link](https://github.com/lowRISC/opentitan/pull/3119#discussion_r473685195).

addressed! please have a look.

rswarbrick reviewed (approved):
Looks great! Thank you for going through all my (zillions of) comments!
